File: src/main/java/mffs/gui/GuiCoercionDeriver.java
package mffs.gui;

import mffs.ModularForceFieldSystem;
import mffs.base.GuiMFFS;
import mffs.base.TileMFFS.TilePacketType;
import mffs.container.ContainerCoercionDeriver;
import mffs.tile.TileCoercionDeriver;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.tileentity.TileEntity;

import org.lwjgl.opengl.GL11;

import universalelectricity.api.UniversalElectricity;
import universalelectricity.api.energy.UnitDisplay;
import universalelectricity.api.energy.UnitDisplay.Unit;
import universalelectricity.api.vector.Vector2;
import calclavia.lib.utility.LanguageUtility;
import cpw.mods.fml.common.network.PacketDispatcher;

public class GuiCoercionDeriver extends GuiMFFS
{
	private TileCoercionDeriver tileEntity;

	public GuiCoercionDeriver(EntityPlayer player, TileCoercionDeriver tileentity)
	{
		super(new ContainerCoercionDeriver(player, tileentity), tileentity);
		this.tileEntity = tileentity;
	}

	@Override
	public void initGui()
	{
		this.textFieldPos = new Vector2(30, 43);
		super.initGui();
		this.buttonList.add(new GuiButton(1, this.width / 2 - 10, this.height / 2 - 28, 58, 20, LanguageUtility.getLocal("gui.deriver.derive")));
	}

	@Override
	protected void drawGuiContainerForegroundLayer(int x, int y)
	{
		this.fontRenderer.drawString(this.tileEntity.getInvName(), this.xSize / 2 - this.fontRenderer.getStringWidth(this.tileEntity.getInvName()) / 2, 6, 4210752);

		this.drawTextWithTooltip("frequency", "%1:", 8, 30, x, y);
		this.textFieldFrequency.drawTextBox();

		GL11.glPushMatrix();
		GL11.glRotatef(-90, 0, 0, 1);
		this.drawTextWithTooltip("upgrade", -95, 140, x, y);
		GL11.glPopMatrix();

		if (this.buttonList.get(1) instanceof GuiButton)
		{
			if (!this.tileEntity.isInversed)
			{
				((GuiButton) this.buttonList.get(1)).displayString = LanguageUtility.getLocal("gui.deriver.derive");
			}
			else
			{
				((GuiButton) this.buttonList.get(1)).displayString = LanguageUtility.getLocal("gui.deriver.integrate");
			}
		}

		this.renderUniversalDisplay(85, 30, this.tileEntity.getWattage(), x, y, Unit.WATT);
		this.fontRenderer.drawString(UnitDisplay.getDisplayShort(UniversalElectricity.DEFAULT_VOLTAGE, Unit.VOLTAGE), 85, 40, 4210752);

		this.drawTextWithTooltip("progress", "%1: " + (this.tileEntity.isActive() ? LanguageUtility.getLocal("gui.deriver.running") : LanguageUtility.getLocal("gui.deriver.idle")), 8, 70, x, y);
		this.drawTextWithTooltip("fortron", "%1: " + UnitDisplay.getDisplayShort(this.tileEntity.getFortronEnergy(), Unit.LITER), 8, 105, x, y);

		this.fontRenderer.drawString((this.tileEntity.isInversed ? "\u00a74-" : "\u00a72+") + UnitDisplay.getDisplayShort(this.tileEntity.getProductionRate() * 20, Unit.LITER) + "/s", 118, 117, 4210752);

		super.drawGuiContainerForegroundLayer(x, y);
	}

	@Override
	protected void drawGuiContainerBackgroundLayer(float f, int x, int y)
	{
		super.drawGuiContainerBackgroundLayer(f, x, y);

		/**
		 * Upgrade Slots
		 */
		this.drawSlot(153, 46);
		this.drawSlot(153, 66);
		this.drawSlot(153, 86);

		/**
		 * Frequency Card Slot
		 */
		this.drawSlot(8, 40);

		/**
		 * Matter Input
		 */
		this.drawSlot(8, 82, SlotType.BATTERY);
		this.drawSlot(8 + 20, 82);

		this.drawBar(50, 84, 1);

		/**
		 * Force Power Bar
		 */
		this.drawForce(8, 115, (float) this.tileEntity.getFortronEnergy() / (float) this.tileEntity.getFortronCapacity());
	}

	@Override
	protected void actionPerformed(GuiButton guibutton)
	{
		super.actionPerformed(guibutton);

		if (guibutton.id == 1)
		{
			PacketDispatcher.sendPacketToServer(ModularForceFieldSystem.PACKET_TILE.getPacket((TileEntity) this.frequencyTile, TilePacketType.TOGGLE_MODE.ordinal()));
		}
	}
}

File: src/main/scala/mffs/production/GuiCoercionDeriver.scala
package mffs.production

import mffs.ModularForceFieldSystem
import mffs.base.{GuiMFFS, TilePacketType}
import net.minecraft.client.gui.GuiButton
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.tileentity.TileEntity
import org.lwjgl.opengl.GL11
import resonant.lib.gui.GuiContainerBase.SlotType
import resonant.lib.network.discriminator.PacketTile
import resonant.lib.render.EnumColor
import resonant.lib.utility.LanguageUtility
import resonant.lib.wrapper.WrapList._
import universalelectricity.api.UnitDisplay

class GuiCoercionDeriver(player: EntityPlayer, tile: TileCoercionDeriver) extends GuiMFFS(new ContainerCoercionDeriver(player, tile), tile)
{
  override def initGui
  {
    super.initGui
    this.buttonList.add(new GuiButton(1, this.width / 2 - 10, this.height / 2 - 35, 58, 20, LanguageUtility.getLocal("gui.deriver.derive")))
  }

  override def drawGuiContainerForegroundLayer(x: Int, y: Int)
  {
    drawStringCentered(tile.getInventoryName)
    GL11.glPushMatrix
    GL11.glRotatef(-90, 0, 0, 1)
    drawTextWithTooltip("upgrade", -95, 140, x, y)
    GL11.glPopMatrix

    if (!tile.isInversed)
    {
      buttonList.get(1).asInstanceOf[GuiButton].displayString = LanguageUtility.getLocal("gui.deriver.derive")
    }
    else
    {
      buttonList.get(1).asInstanceOf[GuiButton].displayString = LanguageUtility.getLocal("gui.deriver.integrate")
    }

    drawString(EnumColor.AQUA + "Energy Requirement:", 8, 20)
    renderUniversalDisplay(8, 30, tile.getPower, x, y, UnitDisplay.Unit.WATT)
    drawString(new UnitDisplay(UnitDisplay.Unit.VOLTAGE, tile.getVoltage).simple().toString, 8, 40)

    drawTextWithTooltip("progress", "%1: " + (if (this.tile.isActive) LanguageUtility.getLocal("gui.deriver.running") else LanguageUtility.getLocal("gui.deriver.idle")), 8, 60, x, y)
    drawString("Production: " + (if (this.tile.isInversed) EnumColor.DARK_RED else EnumColor.DARK_GREEN) + new UnitDisplay(UnitDisplay.Unit.LITER, tile.productionRate * 20) + "/s", 8, 100)

    drawFortronText(x, y)
    super.drawGuiContainerForegroundLayer(x, y)
  }

  override def drawGuiContainerBackgroundLayer(f: Float, x: Int, y: Int)
  {
    super.drawGuiContainerBackgroundLayer(f, x, y)

    //Upgrade slots
    (0 to 2) foreach (y => drawSlot(153, 46 + y * 18))

    drawSlot(8, 75, SlotType.BATTERY)
    drawSlot(8 + 20, 75)
    drawBar(50, 77, 1)

    drawFrequencyGui()
  }

  override def actionPerformed(guibutton: GuiButton)
  {
    super.actionPerformed(guibutton)

    if (guibutton.id == 1)
    {
      ModularForceFieldSystem.packetHandler.sendToServer(new PacketTile(tile, TilePacketType.TOGGLE_MODE.id: Integer))
    }
  }
}